### PR TITLE
Add NAMFileBrowserControl to group Model/IR loading controls

### PR DESCRIPTION
**Pull request by olilarkin**
This PR adds a new control NAMFileBrowserControl, which is used for the Model and IR loading. It

- Groups all controls into a meta control
- Implements browsing through adjacent files in the same directory as the chosen file with left/right arrows
- Implements a pop-up menu with all .nam. or .wav files in the folder that the original file was from
- Eliipsizes long file names that are displayed in the UI
- Enables tooltips, and shows the full filename in the tooltip for the child control that displays the current model/ir name
- Reworks the model-view communication to remove non realtime safe code - the previous approach was calling SendControlMsgFromDelegate() which you shouldn't do on the audio thread

https://user-images.githubusercontent.com/655662/236649896-7ff1ab0b-833c-4cfb-bf83-1cca7cd227ae.mov

### CODE REVIEW

File: NeuralAmpModeler/projects/NeuralAmpModeler-iOS.xcodeproj/project.pbxproj
Comment (sdatkinson) [inline NeuralAmpModeler/projects/NeuralAmpModeler-iOS.xcodeproj/project.pbxproj:249]: Nit: corresponding file additions are missing in the VS solution for Windows

**Review by sdatkinson — CHANGES_REQUESTED** [review]
All of the main changes are 10/10. Awesome to have them! And the metacontrol definition is awesome.

One critical issue that I'm not sure how to solve yet & I'd like to either have a fix or at least a plan before I can merge: The code in [`OnUIOpen`](https://github.com/olilarkin/NeuralAmpModelerPlugin/blob/a564c3594c994960c892be220e371ab957e50c3c/NeuralAmpModeler/NeuralAmpModeler.cpp#L492) doesn't seem to be executing anymore: when I reopen the UI, the model & IR text are back to their defaults "Select model..." etc. It looks like the code that affects what we see is now [`mFileNameButton->SetLabelStr()`](https://github.com/olilarkin/NeuralAmpModelerPlugin/blob/a564c3594c994960c892be220e371ab957e50c3c/NeuralAmpModeler/NeuralAmpModelerControls.h#L178)? Is there a way that this can be accessed from `OnUIOpen`? 

As soon as either there's a fix or I understand how to clean it up, I'm good to merge this.

**Comment by olilarkin** [comment]
> All of the main changes are 10/10. Awesome to have them! And the metacontrol definition is awesome.
> 
> One critical issue that I'm not sure how to solve yet & I'd like to either have a fix or at least a plan before I can merge: The code in [`OnUIOpen`](https://github.com/olilarkin/NeuralAmpModelerPlugin/blob/a564c3594c994960c892be220e371ab957e50c3c/NeuralAmpModeler/NeuralAmpModeler.cpp#L492) doesn't seem to be executing anymore: when I reopen the UI, the model & IR text are back to their defaults "Select model..." etc. It looks like the code that affects what we see is now [`mFileNameButton->SetLabelStr()`](https://github.com/olilarkin/NeuralAmpModelerPlugin/blob/a564c3594c994960c892be220e371ab957e50c3c/NeuralAmpModeler/NeuralAmpModelerControls.h#L178)? Is there a way that this can be accessed from `OnUIOpen`?
> 
> As soon as either there's a fix or I understand how to clean it up, I'm good to merge this.

yeah i am aware that this is not working yet, this PR is a draft it's not quite ready to be merged

**Comment by sdatkinson** [comment]
> this PR is a draft it's not quite ready to be merged

Right, apologies. Still an unintuitive feature to me (drafts that are visible to others) 😅

**Comment by olilarkin** [comment]
Possibly blocked by #254

**Comment by sdatkinson** [comment]
> Possibly blocked by #254

If we want to add the entitlement to `~/Music/NeuralAmpModeler` just to demonstrate that it works, then that's fine and we can pick up the details (eg ensuring its existence) in a follow-up PR.

Just in case that helps manage scope to this PR.

**Comment by James-F2** [comment]
This glorious PR can't come fast enough. 😍
Windows explorer is slow as heck.

**Comment by towner96** [comment]
Thank you! Great advancement. Loading nam files will be faster now.

**Comment by jimc4893** [comment]
is this almost good to go?

File: NeuralAmpModeler/projects/NeuralAmpModeler-iOS.xcodeproj/project.pbxproj
Comment (olilarkin) [reply NeuralAmpModeler/projects/NeuralAmpModeler-iOS.xcodeproj/project.pbxproj:249]: Fixed

File: NeuralAmpModeler/NeuralAmpModeler.cpp
Comment (olilarkin) [reply NeuralAmpModeler/NeuralAmpModeler.cpp]: I intend to do a follow up PR with cleanup for the UI layout, variable naming etc and will clean this up there

**Comment by olilarkin** [comment]
@sdatkinson This is now ready for review. The model and IR path should be updated when the UI opens. There is a tonne of code clean-up around this that I would like to do, but one thing at a time I guess.

**Comment by SoulReaverFan** [comment]
Looking forward to this one! Hope it’s coming in the next update 🙂

**Comment by sdatkinson** [comment]
This is looking great on Windows.

I'm seeing a bug on macOS where the File browser controls aren't getting re-initialized correctly when the UI is re-opened.

Set up: 
image

Close the UI and re-open and:
image

Looking into it.

**Comment by sdatkinson** [comment]
Dropdown, L/R arrows, and tooltip also aren't working on macOS standalone.

**Comment by sdatkinson** [comment]
Deleted the comments with invalid issues.

Please excuse me--I had the wrong commit on my local clone 😅 

Windows is looking good. macOS looking good in the DAW; currently checking the standalone.

**Comment by sdatkinson** [comment]
Update looking into macOS.

There seems to be an issue with loading models and IRs. The behavior is that after clicking "Open" on the file selection dialog, the dialog closes, but the standalone continues with no model loaded.

The model is located at `/Users/steve/Documents/NAM Models/model.nam`.

Putting a breakpoint in `ProcessBlock`, I see that `mStagedNAM` is still `nullptr`.

Looking closer, I've identified that [`loadModelCompletionHandler`]() isn't visited after clicking "Open".

First guess would be that this is related to the entitlements issue Oli called out in a separate Issue.

Continuing to look, will update.

**Review by sdatkinson — APPROVED** [review]
This introduces a bug where laoding NAMs and IRs on the macOS standalone is broken. This is probably because of entitlements and is more or less anticipated in #254.

But we'll merge this and tackle that next.

**Comment by sdatkinson** [comment]
Update/Correction: Seems that Entitlements for the Music directory are already set.

Standalone works if the models are under the Music directory 😄 🎉 

Question is how to inform users about this requirement. It will be a breaking change because users who have stored their models somewhere else will not be able to load them up at all anymore.  Some modal would be great. I'll make an Issue for it.

**Pull request merged**
